src/server/ua_services_method.c

```diff
@@ -178,6 +179,95 @@ callWithMethodAndObject(UA_Server *server, UA_Session *session,
             }
         }
     }
+    if(!found) {
+        /* If the object doesn't have a hasComponent reference to the method node,
+         * check its objectType (and its supertypes). Invoked method can be a component
+         * of objectType and be invoked on this objectType's instance (or on a instance
+         * of one of its subtypes). */
+
+        UA_ReferenceTypeSet hasTypeDefinitionRefs;
+        result->statusCode =
+            referenceTypeIndices(server, &hasTypeDefinitionNodeId,
+                                 &hasTypeDefinitionRefs, true);
+        if(result->statusCode != UA_STATUSCODE_GOOD)
+            return;
+
+        /* Get the ObjectType */
+        const UA_Node *objectType = NULL;
+        bool releaseType = true;
+        if(object->head.nodeClass == UA_NODECLASS_OBJECT) {
+            for(size_t i = 0; i < object->head.referencesSize && !objectType; ++i) {
+                const UA_NodeReferenceKind *rk = &object->head.references[i];
+                if(rk->isInverse)
+                    continue;
+                if(!UA_ReferenceTypeSet_contains(&hasTypeDefinitionRefs, rk->referenceTypeIndex))
+                    continue;
+                /* Node can have only 1 forward HasTypeDefinition reference */
+                const UA_ReferenceTarget *t = NULL;
+                if(!(t = UA_NodeReferenceKind_iterate(rk, t)))
+                    break;
+                objectType = UA_NODESTORE_GETFROMREF(server, t->targetId);
+            }
+        }
+        else {
+            /* If the method was invoked on objectType use the provided node and don't 
+             * release it later, it's done in calling function */
+            objectType = (const UA_Node *)object;
+            releaseType = false;
+        }
+            
+
+        UA_ReferenceTypeSet hasSubTypeRefs;
+        result->statusCode =
+            referenceTypeIndices(server, &hasSubTypeNodeId,
+                                 &hasSubTypeRefs, true);
+        if(result->statusCode != UA_STATUSCODE_GOOD)
+            return;
+
+
+        while(objectType && !found) {
```

Comment:
Use `UA_NodeReferenceKind_findTarget` for this.
This can be much faster for nodes with many children.
Because we directly search in the binary tree instead of iterating.

Reply:
Done as suggested